src/java.base/share/classes/java/util/HexFormat.java

```diff
@@ -0,0 +1,1047 @@
+                    string.length());
+
+        byte[] bytes = new byte[string.length() / 2];
+        int illegal = 0;        // Accumulate logical-or of all bytes
+        for (int i = 0; i < bytes.length; i++) {
+            int v = fromHexDigits(string, i * 2);
+            bytes[i] = (byte) v;
+            illegal |= v;
+        }
+        // check if any character was an illegal character
+        if (illegal < 0)
+            throw new IllegalArgumentException("input contains non-hexadecimal characters");
+
+        return bytes;
+    }
+
+    /**
+     * Check the number of requested digits against a limit.
+     *
+     * @param fromIndex the initial index of the range, inclusive
+     * @param toIndex the final index of the range, exclusive.
+     * @param limit the maximum allowed
+     * @return the length of the range
+     */
+    private static int checkDigitCount(int fromIndex, int toIndex, int limit) {
+        int length = toIndex - fromIndex;
+        if (length > limit)
+            throw new IllegalArgumentException("string length greater than " +
+                    limit + ": " + length);
+        return length;
+    }
+
+    /**
+     * Returns {@code true} if the character is a valid hexadecimal character or codepoint.
+     * The valid hexadecimal characters are:
+     * <ul>
+     * <li>{@code '0' ('\u005Cu0030')} through {@code '9' ('\u005Cu0039')} inclusive,
+     * <li>{@code 'A' ('\u005Cu0041')} through {@code 'F' ('\u005Cu0046')} inclusive, and
+     * <li>{@code 'a' ('\u005Cu0061')} through {@code 'f' ('\u005Cu0066')} inclusive.
+     * </ul>
+     * @param ch a codepoint
+     * @return {@code true} if the character is valid a hexadecimal character,
+     *          otherwise {@code false}
+     */
+    public boolean isHexDigit(int ch) {
+        return ((ch >>> 8) == 0 && DIGITS[ch] >= 0);
+    }
+
+    /**
+     * Returns the value for the hexadecimal character or codepoint.
+     * The value is:
+     * <ul>
+     * <li>{@code (ch - '0')} for {@code '0'} through {@code '9'} inclusive,
+     * <li>{@code (ch - 'A' + 10)} for {@code 'A'} through {@code 'F'} inclusive, and
+     * <li>{@code (ch - 'a' + 10)} for {@code 'a'} through {@code 'f'} inclusive.
+     * </ul>
+     * @param ch a character or codepoint
+     * @return the value {@code 0-15}
+     * @throws  NumberFormatException if the codepoint is not a hexadecimal character
+     */
+    public int fromHexDigit(int ch) {
+        int value;
+        if ((ch >>> 8) == 0 && (value = DIGITS[ch]) >= 0) {
+            return value;
+        }
+        throw new NumberFormatException("not a hexadecimal digit: \"" + (char) ch + "\" = " + ch);
+    }
+
+    /**
+     * Returns a value parsed from two hexadecimal characters in a string.
+     * The characters in the range from {@code index} to {@code index + 1},
+     * inclusive, must be valid hex digits according to {@link #fromHexDigit(int)}.
+     * The delimiter, prefix and suffix are not used.
```

Comment:
In this method - and all similar methods that follow - then is `isUpperCase()` taken into account? If not maybe it should be made explicit - just as for delimiter, prefix, and suffix? 
Alternatively, remove this line and make the method static. Then it would be clear that upperCase, delimiter, prefix, and suffix are not used.

Reply:
This is a private method, documented more than is usual.
The fromHexDigit() method it calls is an instance method to be consistent in the API with the other methods and still have the context of the HexFormat, allowing for possible futures.
(For example, a strict vs lenient setting with respect to uppercase).
The caller is free to use the static method in Character.digit(ch, 16) if that is a better fit..

Reply:
> This is a private method, documented more than is usual.

Sorry - my comment was misplaced. I was mostly concerned with the *public* methods that follow.
(` public int fromHexDigits(CharSequence string)` and friends)
They all say that `* The delimiter, prefix and suffix are not used.` but they do not say anything about upperCase/lowerCase. Since this is normative specification doesn't it need to be fixed?